src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -20,110 +18,128 @@ import com.intellij.openapi.vfs.VirtualFileManager
 import com.intellij.openapi.vfs.newvfs.BulkFileListener
 import com.intellij.openapi.vfs.newvfs.events.VFileEvent
 import com.intellij.util.Alarm
+import com.intellij.util.messages.Topic
 import org.jetbrains.annotations.TestOnly
 import org.rust.cargo.project.CargoProjectDescription
-import org.rust.cargo.project.CargoProjectDescriptionData
 import org.rust.cargo.project.settings.rustSettings
 import org.rust.cargo.project.settings.toolchain
 import org.rust.cargo.project.workspace.CargoProjectWorkspace
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener
+import org.rust.cargo.project.workspace.CargoProjectWorkspaceListener.UpdateResult
 import org.rust.cargo.toolchain.RustToolchain
 import org.rust.cargo.util.cargoLibraryName
 import org.rust.cargo.util.cargoProjectRoot
+import org.rust.cargo.util.enqueue
 import org.rust.cargo.util.updateLibrary
-import kotlin.properties.Delegates
+import org.rust.lang.utils.lock
+import java.util.concurrent.locks.Lock
+import java.util.concurrent.locks.ReentrantLock
 
-private val LOG = Logger.getInstance(CargoProjectWorkspaceImpl::class.java);
 
-@State(
-    name = "CargoMetadata",
-    storages = arrayOf(Storage(file = StoragePathMacros.MODULE_FILE))
-)
-class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorkspace, PersistentStateComponent<CargoProjectState>, BulkFileListener {
+/**
+ * [CargoProjectWorkspace] component implementation
+ */
+class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorkspace, ModuleComponent {
 
-    private val DELAY_MILLIS = 1000
+    private val DELAY = 1000 /* milliseconds */
+
+    private val LOG = Logger.getInstance(CargoProjectWorkspaceImpl::class.java)
 
     /**
      * Alarm used to coalesce consecutive update requests.
-     * It uses EDT thread, but the tasks are really tiny and
-     * only spawn background update.
+     * It uses pooled thread.
      */
     private val alarm = Alarm()
 
-    private var cargoProjectState: CargoProjectState by Delegates.observable(CargoProjectState()) {
-        prop, old, new ->
-            projectDescription = new.projectData?.let { CargoProjectDescription.deserialize(it) }
-    }
+    /**
+     * Lock to guard reads/updates to [cached]
+     */
+    private var lock: Lock = ReentrantLock()
```

Comment:
> Calling arbitrary listeners while holding a lock is asking for trouble.

It's only in the case you can accidentally reverse the order of acquisition of those. That's not the case.

> But I steel don't get it. If you listener does some async work then the lock has no effect. If it does some sync work, then, well, onSucess calls are serialized and you don't need the lock.

w/o the lock you'd abuse your listeners: them would be able to see `projectDescription` prior to `module` instance being updated.